keps/sig-node/4815-dra-partitionable-devices/README.md

```diff
@@ -0,0 +1,2140 @@
+devices:
+- name: gpu-0
+  composite:
+    includes:
+    - name: system-attributes
+    - name: common-gpu-attributes
+    - name: common-gpu-capacities
+    attributes:
+      index:
+        int: 0
+      minor:
+        int: 0
+      uuid:
+        string: GPU-4cf8db2d-06c0-7d70-1a51-e59b25b2c16c
+- name: gpu-1
+  composite:
+    includes:
+    - name: system-attributes
+    - name: common-gpu-attributes
+    - name: common-gpu-capacities
+    attributes:
+      index:
+        int: 1
+      minor:
+        int: 1
+      uuid:
+        string: GPU-4404041a-04cf-1ccf-9e70-f139a9b1e23c
+- name: gpu-2
+  composite:
+    includes:
+    - name: system-attributes
+    - name: common-gpu-attributes
+    - name: common-gpu-capacities
+    attributes:
+      index:
+        int: 2
+      minor:
+        int: 2
+      uuid:
+        string: GPU-79a2ba02-a537-ccbf-2965-8e9d90c0bd54
+- name: gpu-3
+  composite:
+    includes:
+    - name: system-attributes
+    - name: common-gpu-attributes
+    - name: common-gpu-capacities
+    attributes:
+      index:
+        int: 3
+      minor:
+        int: 3
+      uuid:
+        string: GPU-662077db-fa3f-0d8f-9502-21ab0ef058a2
+```
+
+As you can see, three mixins are created called "system-attributes",
+"common-gpu-attributes", and "common-gpu-capacities", which all get included in
+the definitons of the actual GPU devices themselves, along with their own
+device-specific attributes (and device-specific capacities if there were any).
+
+With this in place, the scheduler can parse these device definitions and
+"flatten" them into something that looks exactly the same as a `Basic` device.
+The scheduler can then allocate such devices using the same algorithm it uses
+for `Basic` devices.
+
+### Defining device partitions in terms of consumed capacity in a composite device
+
+A simple example demonstrating how the `ConsumesCapacityFrom` field can be used
+to define multiple, allocatable partitions of a single overarching device can be
+seen below.
+
+```yaml
+devices:
+- name: gpu-0
+  composite:
+    capacity:
+      memory:
+        quantity: 40Gi
+- name: gpu-0-partition-0
+  composite:
+    capacity:
+      memory:
+        quantity: 10Gi
+    consumesCapacityFrom:
+    - name: gpu-0
+- name: gpu-0-partition-1
+  composite:
+    capacity:
+      memory:
+        quantity: 10Gi
+    consumesCapacityFrom:
+    - name: gpu-0
+- name: gpu-0-partition-2
+  composite:
+    capacity:
+      memory:
+        quantity: 10Gi
+    consumesCapacityFrom:
+    - name: gpu-0
+- name: gpu-0-partition-3
+  composite:
+    capacity:
+      memory:
+        quantity: 10Gi
+    consumesCapacityFrom:
+    - name: gpu-0
+```
+
+In this example, five devices are defined: a full GPU called "gpu-0" and four
+partitions of "gpu-0" called "gpu-0-partion-0", "gpu-0-partion-1",
+"gpu-0-partion-2", and "gpu-0-partion-3" respectively. The full GPU has 40Gi of
+memory available, and each of the partitions pull 10Gi from it.
```

Comment:
Remember -- no one is defining these resource slices by hand (the driver publishes these), so the likelihood of a cut-paste-error is minimal. I don't think its worth trying to define semantics that would enable us to recognize that "gpu0 with 40Gi and 4 children each with 20Gi" is an invalid configuration. Driver implementors should be responsible for this themselves. 

> What if the children declare a capacity "foobar" which is not named in the parent? Is that an error?

This would be hard to validate if the parent and child are in different resource slices. I think we can (and should) do our best to validate it in cases where they are in the same slice, but not error out if they are not.

> Is there a need for top-level resources from which children can consume but are not themselves allocatable? e.g. This card has 8 frobnicators and each child device needs 1 or 2, but there's no allocatable device with 8?

Yes, there are use cases for this. In fact, this is the **exact* use case @byako pointed out for SR-IOV VFs. You can achieve this in the current design by defining your `DeviceClass` such that the top level device is never selectable, but its a bit back-handed since this is not encoded anywhere in the `ResourceSlice` API itself.

We toyed with the idea of adding an extra field to a device to accommodate this case (e.g. `abstract: true` or `allocatable: false`), but never made it part of the design. I think the conclusion was that we could rely on `DeviceClasses` for now and think more deeply about this later.

Happy to continue discussing during implementation though.